File: dart/example/consumer_example.dart
// Third-party consumer example. Depends on the INSTALLED selenium package
// (via a pub path dep to the staged copy — NOT the repo source) and proves the
// bundled engine .so loads and drives the protocol, with SELENIUM_CORE_LIB unset
// so only the package's own bundled native/ can satisfy the load.
//
// Discovery: a real consumer resolves the installed package's bundled .so via
// its package URI, then pins it with configureNativeLib(). Modes: ffi | discovery
// | live.
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';

import 'package:selenium/selenium.dart';

Future<String?> bundledSoPath() async {
  final uri = await Isolate.resolvePackageUri(
      Uri.parse('package:selenium/selenium.dart'));
  if (uri == null) return null;
  // <pkg>/lib/selenium.dart -> <pkg>/native/libselenium_core.so
  final libDir = File.fromUri(uri).parent; // .../lib
  final pkgRoot = libDir.parent; // pkg root
  final so = File('${pkgRoot.path}/native/libselenium_core.so');
  return so.existsSync() ? so.path : null;
}

void fail(String msg) {
  stderr.writeln('FAIL: $msg');
  exit(1);
}

Future<void> configureBundled() async {
  final so = await bundledSoPath();
  if (so == null) fail('bundled native/libselenium_core.so not found in the installed package');
  WebDriver.configureNativeLib(so!);
}

Future<void> modeFfi() async {
  await configureBundled();
  if (route('get') != 'POST /session/:sessionId/url') fail('route mismatch');
  if (errorCode('no such element') != 17) fail('errorCode mismatch');
  final byId = By.id('main');
  final loc = jsonDecode(locator(byId.strategy, byId.value)) as Map;
  if (loc['value'] != '*[id="main"]') fail('locator mismatch: $loc');
  try {
    WebDriver.chrome('http://127.0.0.1:1');
    fail('expected transport failure');
  } on WebDriverException catch (e) {
    if (e.code != -1) fail('wrong transport code ${e.code}');
  }
  print('consumer(ffi): OK — installed package loaded its bundled .so and marshalled');
}

Future<void> modeDiscovery() async {
  final env = Platform.environment['SELENIUM_CORE_LIB'];
  if (env != null && env.isNotEmpty) fail('SELENIUM_CORE_LIB set; discovery must run without it');
  final so = await bundledSoPath();
  if (so == null) fail('no bundled .so discovered in the installed package');
  WebDriver.configureNativeLib(so!);
  if (route('newSession') != 'POST /session') fail('route mismatch (bundled .so did not load)');
  print('consumer(discovery): OK — zero-config bundled-.so discovery works');
}

String? which(String cmd) {
  final path = Platform.environment['PATH'] ?? '';
  for (final dir in path.split(':')) {
    final f = File('$dir/$cmd');
    if (f.existsSync()) return f.path;
  }
  return null;
}

Future<int> freePort() async {
  final s = await ServerSocket.bind(InternetAddress.loopbackIPv4, 0);
  final p = s.port;
  await s.close();
  return p;
}

Future<bool> waitUp(int port) async {
  final deadline = DateTime.now().add(const Duration(seconds: 10));
  while (DateTime.now().isBefore(deadline)) {
    try {
      final s = await Socket.connect('127.0.0.1', port,
          timeout: const Duration(milliseconds: 500));
      await s.close();
      return true;
    } catch (_) {
      await Future<void>.delayed(const Duration(milliseconds: 100));
    }
  }
  return false;
}

Future<void> modeLive() async {
  final driver = which('chromedriver');
  if (driver == null) {
    print('consumer(live): SKIPPED — chromedriver not on PATH');
    return;
  }
  await configureBundled();
  final port = await freePort();
  final cd = await Process.start(driver, ['--port=$port'],
      mode: ProcessStartMode.detachedWithStdio);
  try {
    if (!await waitUp(port)) {
      print('consumer(live): SKIPPED — chromedriver did not come up');
      return;
    }
    final d = WebDriver.headlessChrome('http://127.0.0.1:$port');
    try {
      const html = '<!doctype html><title>Installed</title><h1 id="h">Hi</h1>';
      d.get('data:text/html;charset=utf-8,${Uri.encodeComponent(html)}');
      if (d.title != 'Installed') fail('title=${d.title}');
      if (d.findElement(By.id('h')).text != 'Hi') fail('text mismatch');
      print('consumer(live): OK — installed package drove real headless Chrome');
    } finally {
      d.quit();
    }
  } finally {
    cd.kill();
  }
}

Future<void> main(List<String> args) async {
  final mode = args.isNotEmpty ? args[0] : 'ffi';
  switch (mode) {
    case 'ffi':
      await modeFfi();
      break;
    case 'discovery':
      await modeDiscovery();
      break;
    case 'live':
      await modeLive();
      break;
    default:
      fail('unknown mode: $mode');
  }
  // Exit explicitly: a killed chromedriver child can otherwise keep the Dart
  // process alive on its attached stdio, hanging after a successful run.
  exit(0);
}

File: java/example/ConsumerExample.java
import org.openqa.selenium.By;
import org.openqa.selenium.RemoteWebDriver;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.WebDriver;

import java.io.File;
import java.net.Socket;

/**
 * Third-party consumer example. Uses the INSTALLED selenium-core JAR (on the
 * classpath — NOT the source tree) and proves the bundled engine .so (a
 * /native/ jar resource) loads and drives the protocol, with SELENIUM_CORE_LIB
 * unset so only the jar's own bundled .so can satisfy the load.
 *
 * Modes (argv[0]): ffi | discovery | live.
 */
public class ConsumerExample {

    public static void main(String[] args) throws Exception {
        String mode = args.length > 0 ? args[0] : "ffi";
        switch (mode) {
            case "ffi" -> modeFfi();
            case "discovery" -> modeDiscovery();
            case "live" -> modeLive();
            default -> fail("unknown mode: " + mode);
        }
    }

    static void modeFfi() {
        if (!RemoteWebDriver.route("get").equals("POST /session/:sessionId/url")) {
            fail("route mismatch");
        }
        if (RemoteWebDriver.errorCode("no such element") != 17) {
            fail("errorCode mismatch");
        }
        if (!RemoteWebDriver.locator("id", "main").contains("*[id=")) {
            fail("locator mismatch");
        }
        boolean threw = false;
        try {
            RemoteWebDriver.chrome("http://127.0.0.1:1", null);
        } catch (WebDriverException e) {
            threw = e.code() == -1;
        }
        if (!threw) {
            fail("expected transport failure");
        }
        System.out.println("consumer(ffi): OK — installed jar loaded its bundled .so and marshalled");
    }

    static void modeDiscovery() {
        if (System.getenv("SELENIUM_CORE_LIB") != null && !System.getenv("SELENIUM_CORE_LIB").isEmpty()) {
            fail("SELENIUM_CORE_LIB set; discovery must run without it");
        }
        // The bundled .so is a jar resource; if it weren't present, the first
        // native call would fail to load. This call forces the load.
        if (!RemoteWebDriver.route("newSession").equals("POST /session")) {
            fail("route mismatch (bundled .so did not load)");
        }
        System.out.println("consumer(discovery): OK — zero-config bundled-.so (jar resource) discovery works");
    }

    static void modeLive() throws Exception {
        String driver = which("chromedriver");
        if (driver == null) {
            System.out.println("consumer(live): SKIPPED — chromedriver not on PATH");
            return;
        }
        int port;
        try (java.net.ServerSocket s = new java.net.ServerSocket(0, 0, java.net.InetAddress.getByName("127.0.0.1"))) {
            port = s.getLocalPort();
        }
        Process cd = new ProcessBuilder(driver, "--port=" + port)
                .redirectOutput(ProcessBuilder.Redirect.DISCARD)
                .redirectError(ProcessBuilder.Redirect.DISCARD)
                .start();
        try {
            if (!waitUp(port, 10000)) {
                System.out.println("consumer(live): SKIPPED — chromedriver did not come up");
                return;
            }
            WebDriver d = RemoteWebDriver.headlessChrome("http://127.0.0.1:" + port);
            try {
                String html = "<!doctype html><title>Installed</title><h1 id=\"h\">Hi</h1>";
                d.get("data:text/html;charset=utf-8," + java.net.URLEncoder.encode(html, java.nio.charset.StandardCharsets.UTF_8)
                        .replace("+", "%20"));
                if (!d.getTitle().equals("Installed")) {
                    fail("title=" + d.getTitle());
                }
                if (!d.findElement(By.id("h")).getText().equals("Hi")) {
                    fail("text mismatch");
                }
                System.out.println("consumer(live): OK — installed jar drove real headless Chrome");
            } finally {
                d.quit();
            }
        } finally {
            cd.destroy();
        }
    }

    static void fail(String msg) {
        System.err.println("FAIL: " + msg);
        System.exit(1);
    }

    static String which(String cmd) {
        String path = System.getenv("PATH");
        if (path == null) {
            return null;
        }
        for (String dir : path.split(":")) {
            File f = new File(dir, cmd);
            if (f.canExecute() && !f.isDirectory()) {
                return f.getAbsolutePath();
            }
        }
        return null;
    }

    static boolean waitUp(int port, long timeoutMs) {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            try (Socket s = new Socket("127.0.0.1", port)) {
                return true;
            } catch (Exception e) {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException ignored) {
                    Thread.currentThread().interrupt();
                }
            }
        }
        return false;
    }
}
